### PR TITLE
ARROW-9843: [C++] Implement Between trinary kernel

### PR DESCRIPTION
**Pull request by bkmgit**

### CODE REVIEW

**Comment by bkmgit** [comment]
@github-actions crossbow submit -g cpp

**Comment by bkmgit** [comment]
@github-actions crossbow submit -g cpp

**Comment by bkmgit** [comment]
@github-actions crossbow submit -g test-conda-cpp-valgrind

**Comment by bkmgit** [comment]
@github-actions crossbow submit -g cpp-thread-sanitizer

**Comment by bkmgit** [comment]
@github-actions crossbow submit -g cpp

**Comment by bkmgit** [comment]
@github-actions crossbow submit -g cpp

**Comment by bkmgit** [comment]
@github-actions crossbow submit test-conda-cpp-valgrind

**Comment by bkmgit** [comment]
@github-actions crossbow submit test-conda-cpp-valgrind

**Comment by bkmgit** [comment]
@kou Initial review would be helpful. Expect to add `FunctionOptions` as discussed in https://issues.apache.org/jira/browse/ARROW-9843

**Comment by kou** [comment]
Could you create a new Jira issue to handle options? Then we can focus on non-options version in this pull request.

File: cpp/src/arrow/CMakeLists.txt [hunk 1]
```diff
@@ -380,7 +380,7 @@ if(ARROW_COMPUTE)
        compute/exec/hash_join.cc
        compute/exec/hash_join_dict.cc
        compute/exec/hash_join_node.cc
-       compute/exec/ir_consumer.cc
+       compute/exec/ir_consumer.cc
```
Comment (kou) [inline cpp/src/arrow/CMakeLists.txt:383]: Could you revert a needless change?

File: cpp/src/arrow/compute/api_scalar.cc [hunk 2]
```diff
@@ -660,3 +668,4 @@ Result<Datum> Strftime(const Datum& arg, StrftimeOptions options, ExecContext* c
 
 }  // namespace compute
 }  // namespace arrow
+
```
Comment (kou) [inline cpp/src/arrow/compute/api_scalar.cc:671]: Could you revert a needless change?

File: cpp/src/arrow/compute/api_scalar.h [hunk 3]
```diff
@@ -710,6 +710,22 @@ Result<Datum> MinElementWise(
     ElementWiseAggregateOptions options = ElementWiseAggregateOptions::Defaults(),
     ExecContext* ctx = NULLPTR);
 
+/// \brief Between compares each element in `values`
+/// with `left` as a lower bound and 'right' as an upperbound
```
Comment (kou) [inline cpp/src/arrow/compute/api_scalar.h:714]:
```suggestion
/// with `left` as a lower bound and 'right' as an upper bound
```

File: cpp/src/arrow/compute/api_scalar.h [hunk 4]
```diff
@@ -1229,3 +1245,4 @@ ARROW_EXPORT Result<Datum> AssumeTimezone(const Datum& values,
 
 }  // namespace compute
 }  // namespace arrow
+
```
Comment (kou) [inline cpp/src/arrow/compute/api_scalar.h:1248]: Could you revert a needless change?

File: cpp/src/arrow/compute/kernels/codegen_internal.h [hunk 5]
```diff
@@ -445,6 +445,30 @@ static void VisitTwoArrayValuesInline(const ArrayData& arr0, const ArrayData& ar
                         arr0.length, std::move(visit_valid), std::move(visit_null));
 }
 
+template <typename Arg0Type, typename Arg1Type, typename Arg2Type,
+	 typename VisitFunc, typename NullFunc>
+static void VisitThreeArrayValuesInline(const ArrayData& arr0, const ArrayData& arr1,
+                                      const ArrayData& arr2,
+				      VisitFunc&& valid_func, NullFunc&& null_func) {
```
Comment (kou) [inline cpp/src/arrow/compute/kernels/codegen_internal.h:452]: Is this indent accepted by lint?

Could you remove "WIP" from the pull request title?

File: cpp/src/arrow/compute/api_scalar.cc (hunk 2 above)
Comment (bkmgit) [reply cpp/src/arrow/compute/api_scalar.cc:671]: ok

File: cpp/src/arrow/CMakeLists.txt (hunk 1 above)
Comment (bkmgit) [reply cpp/src/arrow/CMakeLists.txt:383]: ok

Comment (bkmgit) [reply cpp/src/arrow/CMakeLists.txt]: fixed formatting error

File: cpp/src/arrow/compute/api_scalar.cc (hunk 2 above)
Comment (bkmgit) [reply cpp/src/arrow/compute/api_scalar.cc]: fixed formatting error

File: cpp/src/arrow/compute/api_scalar.h (hunk 4 above)
Comment (bkmgit) [reply cpp/src/arrow/compute/api_scalar.h]: fixed formatting error

**Comment by kou** [comment]
Could you rebase on master to remove the changes in #11785?

**Comment by kou** [comment]
There are many test failures. It seems that merging is failed.

**Comment by bkmgit** [comment]
Yes, will need to redo this and do local tests first.

**Comment by kou** [comment]
Close in favor of #11882.

**Pull request closed without merging**